### PR TITLE
Slicing IO read_ routines

### LINKED ISSUE #584: Slicing read_* methods
In my opinion, `read_*` routines -where `*` is a word in singular- should not worry about skipping and they should just focus on reading the next thing available. Then we could get rid of all these extra arguments, which will be very hard to document in a way that doesn't confuse users, specially the different defaults for different methods/siles.

It seems clean and clear enough for me to specify the iterator externally (slicing, `multi`, `slice`,`apply`...), just as it is done with the BZ methods for example. If some method is very common to use with `all=True`, there could be an extra method with the plural word, i.e. `read_energy` -> `read_energies` as an alternative for `read_energy[:]`.

_Originally posted by @pfebrer in https://github.com/zerothi/sisl/issues/578#issuecomment-1591358548_

### PR DESCRIPTION
**Pull request by zerothi**
- [x] Closes #584
 - [x] Tests added
 - [x] Documentation for functionality
     somewhat, more documentation would be great.
 - [x] User visible changes (including notable bug fixes) are documented in `CHANGELOG`

### CODE REVIEW

File: src/sisl/io/_multiple.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        # this makes it work like a function bound to an instance (func.__self__
+        # works for instances)
+        self.__self__ = obj
+        self.__func__ = func
+        self.key = key
+        if skip_func is None:
+            self.skip_func = func
+        else:
+            self.skip_func = skip_func
+        self.postprocess = postprocess
```
Comment (tfrederiksen) [inline src/sisl/io/_multiple.py:41]: What happens if `postprocess is None`? Could `list` be a default?

File: src/sisl/io/siesta/tests/test_ani.py [hunk 2]
```diff
@@ -32,41 +32,41 @@ def test_ani(sisl_tmp):
...
-    g = a.read_geometry(start=1, all=False)
+    g = a.read_geometry[1]()
     assert g.na == 2
 
-    g = a.read_geometry(start=1, stop=3, step=1)
+    g = a.read_geometry[1:3:1]()
     assert g[1].na == 3 and len(g) == 2
 
-    g = a.read_geometry(start=1, stop=3, step=2, all=True)
+    g = a.read_geometry[1:3:2]()
```
Comment (tfrederiksen) [inline src/sisl/io/siesta/tests/test_ani.py:69]: What about adding some checks with negative indices?

**Review by tfrederiksen — COMMENTED** [review]
I played a bit around with this and it looks really cool. Well done!

File: src/sisl/io/_multiple.py (hunk 1 above)
Comment (zerothi) [reply src/sisl/io/_multiple.py:41]: It then returns as the decorated routine returns the values (in this case a list of the return values).
I see there was no default, now it is `lambda x: x` (or the equivalent).

So yes, the default is a list.

File: src/sisl/io/siesta/tests/test_ani.py (hunk 2 above)
Comment (zerothi) [reply src/sisl/io/siesta/tests/test_ani.py:69]: Added a new test. (I'll keep from pushing, I am reaching the bandwidth limit for the lfs repo)

**Comment by zerothi** [comment]
I don't see a reason to not use this everywhere, i.e. read_energy etc?

Would there be objections to this?

**Comment by tfrederiksen** [comment]
> I don't see a reason to not use this everywhere, i.e. read_energy etc?
> 
> Would there be objections to this?

I agree, no objections. I already tested this out on `stdoutSileVASP.read_energy()`. However, I tend to think it may cause confusion if we change the behaviour everywhere to return the *first* entry instead of the *last* (default up to now).

Maybe raise a warning for a while to make users aware that `read_energy[-1]()` is needed to get the last entry (that I suppose is the typical case)?

**Comment by zerothi** [comment]
Where should we default to return the last? Everywhere, some of @pfebrer comments related to MD are valid.
However, I assume that those doing MD are aware of their needs and thus does require to return all.
I don't know if *all* should default to return the last item?

**Comment by pfebrer** [comment]
Wouldn't the default slice return the confusion of the arguments we have removed. I advocate for no defaults on this, then "legacy" code will be more likely to keep doing the same thing in the future. If in the future we decide that another default is better, it will be very difficult to debug changes in the defaults in already written code.

I would just issue a warning for some time as @tfrederiksen was suggesting (perhaps keep the current API with a deprecation warning).

Also in my case I would spend more time rechecking the documentation for the defaults of each sile than simply specifying what I want 😅

**Comment by pfebrer** [comment]
By the way I don't know if it would be a good idea to use slices for cases where there is more than one dimension. E.g. in `outSileSiesta.read_scf` one could do `read_scf[:, -1]` meaning all MD steps, but just the last scf step. 

It looks nice but it's a bit hard to know if it makes sense implementationwise because it would no longer be just a simple external iterator on the reading function.

**Comment by pfebrer** [comment]
> Yeah, you are probably right. But if we then just are choosing a sane default that applies everywhere? There are pros and cons on choosing first/last, which I think are the only viable ones.

😂
My point is no default, and let the read function read at whatever position the file handle is.

**Comment by zerothi** [comment]
> > Yeah, you are probably right. But if we then just are choosing a sane default that applies everywhere? There are pros and cons on choosing first/last, which I think are the only viable ones.
> 
> joy My point is no default, and let the read function read at whatever position the file handle is.

But a no default is still a default (`[0]`). So I think that is still up for discussion. ;)

I am probably pro the last, since anything but MD are generally most interested in the last (say converged system).

**Comment by zerothi** [comment]
> By the way I don't know if it would be a good idea to use slices for cases where there is more than one dimension. E.g. in `outSileSiesta.read_scf` one could do `read_scf[:, -1]` meaning all MD steps, but just the last scf step.
> 
> It looks nice but it's a bit hard to know if it makes sense implementationwise because it would no longer be just a simple external iterator on the reading function.

This is a fine idea, I think it could be implemented easily as the subsequent tuple keys are simply applied to the returned arrays. Lets see...

**Comment by pfebrer** [comment]
> But a no default is still a default ([0]). So I think that is still up for discussion. ;)

Then what would

```python
with sile as f:
    for i in range(3):
        f.read_geometry()
```

do? Read three times the first or the last geometry?

**Comment by zerothi** [comment]
> > But a no default is still a default ([0]). So I think that is still up for discussion. ;)
> 
> Then what would
> 
> ```python
> with sile as f:
>     for i in range(3):
>         f.read_geometry()
> ```
> 
> do? Read three times the first or the last geometry?

ah yeah, there was this annoyance... damn... why is this so annoying...

Would it be too problematic to check if the file was open? Would that be annoying for end-users?

.e.g.
```python
last = f.read_geometry()
# whereas
with f:
     next = f.read_geometry()
```
?

**Comment by pfebrer** [comment]
Would it be so bad to have a `read_last_geometry()` method if it makes sense for some `Sile`?

**Comment by tfrederiksen** [comment]
> Would it be so bad to have a `read_last_geometry()` method if it makes sense for some `Sile`?

I think this is a good suggestion, making it explicit for the user what is looked up. In particular the warning could suggest users to adapt `read_last_energy()` in their scripts.

**Comment by pfebrer** [comment]
> I am pretty sure that would make sense for all siles that uses the SileBinder

Yeah maybe 😅

I guess if `read_*` functions only parse lines it will be also easier to combine them. For example you could implement something to read geometries and energies in the same pass through the file.

**Comment by zerothi** [comment]
I have thought about it, and I have now added this, what do you think?
```python
read_geometry.last(*args, **kwargs)
read_geometry.next(*args, **kwargs)
```
to explicitly request the next item, and the last item. Remark that `first` would have been misleading as in the case of @pfebrer's case above.

**Comment by pfebrer** [comment]
Sounds good!

But what does `read_geometry()` do?

**Comment by zerothi** [comment]
> Sounds good!
> 
> But what does `read_geometry()` do?

same as `next` so I don't know if that is needed? but more for clarity, as you suggested.

**Comment by pfebrer** [comment]
Hmm it is very difficult to know what users would think about it. But it doesn't seem it could be harmful in the future to maintain this two properties (fingers crossed haha) so we could have them both.

**Comment by zerothi** [comment]
@tfrederiksen you might be interested in the latest commit which changes the `orca` output to follow this method. I couldn't really safe the `all` argument, but I hope this is fine. In the end when we get to a release it will be 0.14. Let me know your thoughts.

Probably this should go in ASAP, and then the other siles will come as required.

**Comment by pfebrer** [comment]
Could `read_energy` and similars return a numpy array by default instead of a `Collection`?

**Comment by zerothi** [comment]
> Could `read_energy` and similars return a numpy array by default instead of a `Collection`?

That would be hard for end users, then the users need to read which indices corresponds to which energies, that was why the property dict was used.  
It would be nice if they all allow collection or dataframes.

**Review by tfrederiksen — APPROVED** [review]
I think this looks great. After merge I will revisit #573.

**Pull request merged**